In [118]:
# importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [119]:
# loading data
path = r'C:\Users\Hp\Desktop\Salesdata\Salesdata (1).xlsx'
excel_df = pd.ExcelFile(path)
print(excel_df.sheet_names) #listing all the excel sheets


['Main Data', 'Data Dictionary', 'Product Master']


In [120]:
#loading the sheets
main_data = pd.read_excel(path, sheet_name='Main Data')
data_dictionary = pd.read_excel(path, sheet_name='Data Dictionary')
product_master = pd.read_excel(path, sheet_name='Product Master')

In [121]:
#displaying the dataframes
print("Main Data:")
print(main_data.head(10))

Main Data:
   OrderNumber  QuantityOrdered  PriceEach  OrderLineNumber    Sales   Status  \
0        10100               49      34.47                1  1689.03  Shipped   
1        10100               50      67.80                2  3390.00  Shipped   
2        10100               30     100.00                3  5151.00  Shipped   
3        10100               22      86.51                4  1903.22  Shipped   
4        10101               26     100.00                1  3773.38  Shipped   
5        10101               46      53.76                2  2472.96  Shipped   
6        10101               45      31.20                3  1404.00  Shipped   
7        10101               25     100.00                4  3782.00  Shipped   
8        10102               41      50.14                1  2055.74  Shipped   
9        10102               39     100.00                2  4808.31  Shipped   

   QTR_ID  DAY_ID  MONTH_ID  YEAR_ID  ...              Phone  \
0       1       6         1     2

In [122]:
print(main_data.shape)
print(main_data.columns)

(2823, 22)
Index(['OrderNumber', 'QuantityOrdered', 'PriceEach', 'OrderLineNumber',
       'Sales', 'Status', 'QTR_ID', 'DAY_ID', 'MONTH_ID', 'YEAR_ID',
       'ProductCode', 'CustomerName', 'Phone', 'AddressLine1', 'City',
       'PostalCode', 'Country', 'Territory', 'ContactLastName',
       'ContactFirstName', 'Branch', 'CostPrice'],
      dtype='object')


In [123]:
#product master
print("product master:")
print(product_master.head(10))

product master:
    ProductLine ProductCode  MSRP
0  Classic Cars    S10_1949   214
1  Classic Cars    S10_4757   136
2  Classic Cars    S10_4962   147
3  Classic Cars    S12_1099   194
4  Classic Cars    S12_1108   207
5  Classic Cars    S12_3148   151
6  Classic Cars    S12_3380   117
7  Classic Cars    S12_3891   173
8  Classic Cars    S12_3990    79
9  Classic Cars    S12_4675   115


In [124]:
print(product_master.shape)
print(product_master.columns)

(109, 3)
Index(['ProductLine', 'ProductCode', 'MSRP'], dtype='object')


In [125]:
print("data dictionary:")
print(data_dictionary)

data dictionary:
    Unnamed: 0  Unnamed: 1  Unnamed: 2        Field Name  \
0          NaN         NaN         NaN       OrderNumber   
1          NaN         NaN         NaN   QuantityOrdered   
2          NaN         NaN         NaN         PriceEach   
3          NaN         NaN         NaN   OrderLineNumber   
4          NaN         NaN         NaN             Sales   
5          NaN         NaN         NaN         OrderDate   
6          NaN         NaN         NaN            Status   
7          NaN         NaN         NaN            QTR_ID   
8          NaN         NaN         NaN            DAY_ID   
9          NaN         NaN         NaN          MONTH_ID   
10         NaN         NaN         NaN           YEAR_ID   
11         NaN         NaN         NaN       ProductLine   
12         NaN         NaN         NaN              MSRP   
13         NaN         NaN         NaN       ProductCode   
14         NaN         NaN         NaN      CustomerName   
15         NaN         

The datasets main_data and product_master have one common column **ProductCode** hence we will merge the two on the ProductCode column.
The data dictionary dataset explains the columns in the two datasets :(product_master and main_data)

In [126]:
#merging the datasets on product code
merged_df = pd.merge(main_data, product_master, on='ProductCode', how='inner')
merged_df.head()
merged_df.shape

(2823, 24)

In [127]:
# checking stats summary
merged_df.describe()

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,CostPrice,MSRP
count,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.000000,2823.00000,2823.000000,2823.000000
mean,10258.725115,35.092809,83.658544,6.466171,3553.889072,2.717676,14.291534,7.092455,2022.81509,54.378054,100.715551
std,92.085478,9.741443,20.174277,4.225841,1841.865106,1.203878,8.777409,3.656633,0.69967,13.113280,40.187912
min,10100.000000,6.000000,26.880000,1.000000,482.130000,1.000000,1.000000,1.000000,2022.00000,17.472000,33.000000
25%,10180.000000,27.000000,68.860000,3.000000,2203.430000,2.000000,6.000000,4.000000,2022.00000,44.759000,68.000000
50%,10262.000000,35.000000,95.700000,6.000000,3184.800000,3.000000,14.000000,8.000000,2023.00000,62.205000,99.000000
75%,10333.500000,43.000000,100.000000,9.000000,4508.000000,4.000000,21.000000,11.000000,2023.00000,65.000000,124.000000
max,10425.000000,97.000000,100.000000,18.000000,14082.800000,4.000000,31.000000,12.000000,2024.00000,65.000000,214.000000


In [128]:
#calculation of missing values in %
merged_df.isnull().sum()/len(merged_df) * 100

OrderNumber          0.000000
QuantityOrdered      0.000000
PriceEach            0.000000
OrderLineNumber      0.000000
Sales                0.000000
Status               0.000000
QTR_ID               0.000000
DAY_ID               0.000000
MONTH_ID             0.000000
YEAR_ID              0.000000
ProductCode          0.000000
CustomerName         0.000000
Phone                0.000000
AddressLine1         0.000000
City                 0.000000
PostalCode           2.692171
Country              0.000000
Territory           38.044633
ContactLastName      0.000000
ContactFirstName     0.000000
Branch               0.000000
CostPrice            0.000000
ProductLine          0.000000
MSRP                 0.000000
dtype: float64

Only two columns have missing values.PostalCode has almost 3% of the values missing and Territory has 38% of its values missing

**Handling the Territory column**

In [129]:
merged_df['Territory'].value_counts()

Territory
EMEA     1407
APAC      221
Japan     121
Name: count, dtype: int64

EMEA: 1407: This indicates that there are 1407 entries in your dataset where the Territory is labeled as EMEA (Europe, Middle East, and Africa). It's the most frequently occurring value.

APAC: 221: This shows that there are 221 entries with the Territory labeled as APAC (Asia-Pacific). It's the second most common value.

Japan: 121: This indicates that 121 entries have the Territory labeled as Japan.

In [130]:
# rows where territory is missing
missing_territory = merged_df.loc[merged_df['Territory'].isnull()]
missing_territory

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,City,PostalCode,Country,Territory,ContactLastName,ContactFirstName,Branch,CostPrice,ProductLine,MSRP
0,10100,49,34.47,1,1689.03,Shipped,1,6,1,2022,...,Nashua,62005,USA,NaN,Young,Valarie,Skates Road,22.4055,Vintage Cars,41
2,10124,46,33.23,4,1528.58,Shipped,2,21,5,2022,...,Las Vegas,83030,USA,NaN,King,Sue,Skates Road,21.5995,Vintage Cars,41
3,10149,26,38.98,9,1013.48,Shipped,3,12,9,2022,...,Brisbane,94217,USA,NaN,Taylor,Sue,Skates Road,25.3370,Vintage Cars,41
4,10162,37,38.98,7,1442.26,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NaN,Brown,Julie,Skates Road,25.3370,Vintage Cars,41
6,10182,23,42.26,8,971.98,Shipped,4,12,11,2022,...,San Rafael,97562,USA,NaN,Nelson,Valarie,Skates Road,27.4690,Vintage Cars,41
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2813,10278,23,100.00,2,2604.52,Shipped,3,6,8,2023,...,Las Vegas,83030,USA,NaN,King,Sue,Skates Road,65.0000,Classic Cars,124
2816,10321,44,100.00,6,4489.76,Shipped,4,4,11,2023,...,New Bedford,50553,USA,NaN,Benitez,Violeta,Bergenville,65.0000,Classic Cars,124
2817,10331,44,100.00,14,4849.24,Shipped,4,17,11,2023,...,Philadelphia,71270,USA,NaN,Hernandez,Rosa,Bergenville,65.0000,Classic Cars,124
2819,10367,49,56.30,1,2758.70,Resolved,1,12,1,2024,...,Pasadena,90003,USA,NaN,Young,Julie,Skates Road,36.5950,Classic Cars,124


In [131]:
merged_df['Country'].value_counts()

Country
USA            1004
Spain           342
France          314
Australia       185
UK              144
Italy           113
Finland          92
Norway           85
Singapore        79
Canada           70
Denmark          63
Germany          62
Sweden           57
Austria          55
Japan            52
Belgium          33
Switzerland      31
Philippines      26
Ireland          16
Name: count, dtype: int64

In [132]:
#Filling rows where territory is missing and country is USA with NA(North America)
merged_df.loc[(merged_df['Territory'].isnull()) & (merged_df['Country'] =='USA'), 'Territory'] = 'NA'
merged_df.loc[merged_df['Country'] == 'USA'].head(10)

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,City,PostalCode,Country,Territory,ContactLastName,ContactFirstName,Branch,CostPrice,ProductLine,MSRP
0,10100,49,34.47,1,1689.03,Shipped,1,6,1,2022,...,Nashua,62005,USA,NA,Young,Valarie,Skates Road,22.4055,Vintage Cars,41
2,10124,46,33.23,4,1528.58,Shipped,2,21,5,2022,...,Las Vegas,83030,USA,NA,King,Sue,Skates Road,21.5995,Vintage Cars,41
3,10149,26,38.98,9,1013.48,Shipped,3,12,9,2022,...,Brisbane,94217,USA,NA,Taylor,Sue,Skates Road,25.3370,Vintage Cars,41
4,10162,37,38.98,7,1442.26,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,25.3370,Vintage Cars,41
6,10182,23,42.26,8,971.98,Shipped,4,12,11,2022,...,San Rafael,97562,USA,NA,Nelson,Valarie,Skates Road,27.4690,Vintage Cars,41
8,10204,39,33.23,2,1295.97,Shipped,4,2,12,2022,...,NYC,10022,USA,NA,Young,Jeff,Skates Road,21.5995,Vintage Cars,41
11,10242,46,36.93,1,1698.78,Shipped,2,20,4,2023,...,NYC,10022,USA,NA,Kuo,Kee,Skates Road,24.0045,Vintage Cars,41
15,10312,31,35.29,15,1093.99,Shipped,4,21,10,2023,...,San Rafael,97562,USA,NA,Nelson,Valarie,Skates Road,22.9385,Vintage Cars,41
17,10346,22,97.44,4,2143.68,Shipped,4,29,11,2023,...,Las Vegas,83030,USA,NA,King,Sue,Skates Road,63.3360,Vintage Cars,41
18,10368,46,37.34,3,1717.64,Shipped,1,19,1,2024,...,San Rafael,97562,USA,NA,Nelson,Valarie,Skates Road,24.2710,Vintage Cars,41


In [133]:
merged_df.isnull().sum()/len(merged_df) * 100

OrderNumber         0.000000
QuantityOrdered     0.000000
PriceEach           0.000000
OrderLineNumber     0.000000
Sales               0.000000
Status              0.000000
QTR_ID              0.000000
DAY_ID              0.000000
MONTH_ID            0.000000
YEAR_ID             0.000000
ProductCode         0.000000
CustomerName        0.000000
Phone               0.000000
AddressLine1        0.000000
City                0.000000
PostalCode          2.692171
Country             0.000000
Territory           2.479632
ContactLastName     0.000000
ContactFirstName    0.000000
Branch              0.000000
CostPrice           0.000000
ProductLine         0.000000
MSRP                0.000000
dtype: float64

After filling the missing values in the territory column where country is USA with NA(North America) the % of missing values in the territory column reduced to 2%

In [134]:
missing_territory_2 = merged_df.loc[merged_df['Territory'].isnull()]
missing_territory_2

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,City,PostalCode,Country,Territory,ContactLastName,ContactFirstName,Branch,CostPrice,ProductLine,MSRP
259,10313,42,100.00,4,5581.80,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NaN,Tannamuri,Yoshi,Bergenville,65.0000,Trucks and Buses,127
286,10313,34,52.87,5,1797.58,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NaN,Tannamuri,Yoshi,Skates Road,34.3655,Trucks and Buses,60
314,10313,27,87.64,6,2366.28,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NaN,Tannamuri,Yoshi,Skates Road,56.9660,Trucks and Buses,96
321,10411,26,100.00,1,2904.72,Shipped,2,1,5,2024,...,Montreal,H1J 1C3,Canada,NaN,Fresnisre,Jean,Skates Road,65.0000,Trucks and Buses,96
341,10313,40,100.00,7,6678.00,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NaN,Tannamuri,Yoshi,Bergenville,65.0000,Classic Cars,147
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1887,10283,21,98.06,4,2059.26,Shipped,3,20,8,2023,...,Tsawassen,T2F 8M4,Canada,NaN,Lincoln,Elizabeth,Skates Road,63.7390,Ships,86
2089,10235,41,100.00,1,4177.49,Shipped,2,2,4,2023,...,Tsawassen,T2F 8M4,Canada,NaN,Lincoln,Elizabeth,Bergenville,65.0000,Motorcycles,99
2563,10171,35,100.00,1,4680.20,Shipped,4,5,11,2022,...,Montreal,H1J 1C3,Canada,NaN,Fresnisre,Jean,Bergenville,65.0000,Classic Cars,142
2590,10171,35,100.00,2,4508.00,Shipped,4,5,11,2022,...,Montreal,H1J 1C3,Canada,NaN,Fresnisre,Jean,Bergenville,65.0000,Classic Cars,141


since Canada is also in North America we will fill territory where country is canada with NA

In [135]:
#Filling rows where territory is missing and country is Canada with NA(North America)
merged_df.loc[(merged_df['Territory'].isnull()) & (merged_df['Country'] == 'Canada'), 'Territory'] = 'NA'
merged_df.loc[merged_df['Country'] == 'Canada'].head(10)

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,City,PostalCode,Country,Territory,ContactLastName,ContactFirstName,Branch,CostPrice,ProductLine,MSRP
259,10313,42,100.00,4,5581.80,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NA,Tannamuri,Yoshi,Bergenville,65.0000,Trucks and Buses,127
286,10313,34,52.87,5,1797.58,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NA,Tannamuri,Yoshi,Skates Road,34.3655,Trucks and Buses,60
314,10313,27,87.64,6,2366.28,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NA,Tannamuri,Yoshi,Skates Road,56.9660,Trucks and Buses,96
321,10411,26,100.00,1,2904.72,Shipped,2,1,5,2024,...,Montreal,H1J 1C3,Canada,NA,Fresnisre,Jean,Skates Road,65.0000,Trucks and Buses,96
341,10313,40,100.00,7,6678.00,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NA,Tannamuri,Yoshi,Bergenville,65.0000,Classic Cars,147
349,10411,27,100.00,2,4427.73,Shipped,2,1,5,2024,...,Montreal,H1J 1C3,Canada,NA,Fresnisre,Jean,Bergenville,65.0000,Classic Cars,147
369,10313,28,100.00,8,2881.76,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NA,Tannamuri,Yoshi,Skates Road,65.0000,Trucks and Buses,121
376,10411,46,100.00,3,5235.72,Shipped,2,1,5,2024,...,Montreal,H1J 1C3,Canada,NA,Fresnisre,Jean,Bergenville,65.0000,Trucks and Buses,121
387,10206,33,100.00,1,3871.89,Shipped,4,5,12,2022,...,Vancouver,V3F 2K1,Canada,NA,Tannamuri,Yoshi,Bergenville,65.0000,Classic Cars,101
396,10313,30,99.13,9,2973.90,Shipped,4,22,10,2023,...,Vancouver,V3F 2K1,Canada,NA,Tannamuri,Yoshi,Skates Road,64.4345,Classic Cars,101


In [136]:
merged_df['Territory'].value_counts()

Territory
EMEA     1407
NA       1074
APAC      221
Japan     121
Name: count, dtype: int64

In [137]:
merged_df.isnull().sum()/len(merged_df) * 100

OrderNumber         0.000000
QuantityOrdered     0.000000
PriceEach           0.000000
OrderLineNumber     0.000000
Sales               0.000000
Status              0.000000
QTR_ID              0.000000
DAY_ID              0.000000
MONTH_ID            0.000000
YEAR_ID             0.000000
ProductCode         0.000000
CustomerName        0.000000
Phone               0.000000
AddressLine1        0.000000
City                0.000000
PostalCode          2.692171
Country             0.000000
Territory           0.000000
ContactLastName     0.000000
ContactFirstName    0.000000
Branch              0.000000
CostPrice           0.000000
ProductLine         0.000000
MSRP                0.000000
dtype: float64

The territory part is clean and has no missing values


**Handling the PostalCode column**

In [138]:
merged_df['PostalCode'].value_counts()

PostalCode
28034      259
97562      205
10022      152
94217       89
50553       61
          ... 
41101       15
80686       14
WA1 1DP     12
B-6000       8
92561        3
Name: count, Length: 73, dtype: int64

In [139]:
merged_df

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,City,PostalCode,Country,Territory,ContactLastName,ContactFirstName,Branch,CostPrice,ProductLine,MSRP
0,10100,49,34.47,1,1689.03,Shipped,1,6,1,2022,...,Nashua,62005,USA,NA,Young,Valarie,Skates Road,22.4055,Vintage Cars,41
1,10110,48,34.47,5,1654.56,Shipped,1,18,3,2022,...,Manchester,EC2 5NT,UK,EMEA,Ashworth,Victoria,Skates Road,22.4055,Vintage Cars,41
2,10124,46,33.23,4,1528.58,Shipped,2,21,5,2022,...,Las Vegas,83030,USA,NA,King,Sue,Skates Road,21.5995,Vintage Cars,41
3,10149,26,38.98,9,1013.48,Shipped,3,12,9,2022,...,Brisbane,94217,USA,NA,Taylor,Sue,Skates Road,25.3370,Vintage Cars,41
4,10162,37,38.98,7,1442.26,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,25.3370,Vintage Cars,41
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2818,10343,36,100.00,4,5848.92,Shipped,4,24,11,2023,...,Reims,51100,France,EMEA,Henriot,Paul,Bergenville,65.0000,Classic Cars,124
2819,10367,49,56.30,1,2758.70,Resolved,1,12,1,2024,...,Pasadena,90003,USA,NA,Young,Julie,Skates Road,36.5950,Classic Cars,124
2820,10378,34,42.64,5,1449.76,Shipped,1,10,2,2024,...,Madrid,28034,Spain,EMEA,Freyre,Diego,Skates Road,27.7160,Classic Cars,124
2821,10407,59,100.00,11,7048.14,On Hold,2,22,4,2024,...,San Jose,94217,USA,NA,Frick,Sue,Geiselweg,65.0000,Classic Cars,124


In [140]:
# displaying the  rows where postalCode has missing values
missing_postalcode = merged_df.loc[merged_df['PostalCode'].isnull()]
missing_postalcode

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,City,PostalCode,Country,Territory,ContactLastName,ContactFirstName,Branch,CostPrice,ProductLine,MSRP
4,10162,37,38.98,7,1442.26,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,25.337,Vintage Cars,41
26,10162,27,69.62,8,1879.74,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,45.253,Vintage Cars,60
49,10162,29,100.00,9,5176.50,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Bergenville,65.000,Vintage Cars,170
71,10162,39,100.00,10,3912.09,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Bergenville,65.000,Vintage Cars,92
93,10162,48,100.00,3,7209.12,Shipped,4,18,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Geiselweg,65.000,Vintage Cars,168
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2411,10159,35,35.40,9,1239.00,Shipped,4,10,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,23.010,Motorcycles,40
2436,10159,31,71.60,10,2219.60,Shipped,4,10,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,46.540,Motorcycles,81
2461,10159,27,80.34,11,2169.18,Shipped,4,10,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,52.221,Motorcycles,69
2486,10159,23,100.00,12,2347.15,Shipped,4,10,10,2022,...,San Francisco,NaN,USA,NA,Brown,Julie,Skates Road,65.000,Motorcycles,102


In [141]:
merged_df.columns

Index(['OrderNumber', 'QuantityOrdered', 'PriceEach', 'OrderLineNumber',
       'Sales', 'Status', 'QTR_ID', 'DAY_ID', 'MONTH_ID', 'YEAR_ID',
       'ProductCode', 'CustomerName', 'Phone', 'AddressLine1', 'City',
       'PostalCode', 'Country', 'Territory', 'ContactLastName',
       'ContactFirstName', 'Branch', 'CostPrice', 'ProductLine', 'MSRP'],
      dtype='object')

In [142]:
#displaying city,postalcode and adressline columns where postalCode is NAN
missing_postalcode_details = merged_df.loc[merged_df['PostalCode'].isnull(),['City','PostalCode','AddressLine1']]
missing_postalcode_details

,City,PostalCode,AddressLine1
4,San Francisco,NaN,7734 Strong St.
26,San Francisco,NaN,7734 Strong St.
49,San Francisco,NaN,7734 Strong St.
71,San Francisco,NaN,7734 Strong St.
93,San Francisco,NaN,7734 Strong St.
...,...,...,...
2411,San Francisco,NaN,7734 Strong St.
2436,San Francisco,NaN,7734 Strong St.
2461,San Francisco,NaN,7734 Strong St.
2486,San Francisco,NaN,7734 Strong St.


In [143]:
# Filter rows where City is 'San Francisco' and display specific columns
san_francisco_details = merged_df.loc[merged_df['City'] == 'San Francisco', ['City', 'PostalCode', 'AddressLine1']]

# Display the result
san_francisco_details


,City,PostalCode,AddressLine1
4,San Francisco,NaN,7734 Strong St.
26,San Francisco,NaN,7734 Strong St.
49,San Francisco,NaN,7734 Strong St.
71,San Francisco,NaN,7734 Strong St.
93,San Francisco,NaN,7734 Strong St.
...,...,...,...
2386,San Francisco,NaN,7734 Strong St.
2411,San Francisco,NaN,7734 Strong St.
2436,San Francisco,NaN,7734 Strong St.
2461,San Francisco,NaN,7734 Strong St.


In [144]:
# filling the missing postalcode values for San Francisco
merged_df.loc[(merged_df['City'] == 'San Francisco') & (merged_df['PostalCode'].isnull()), 'PostalCode'] = '94101'
# filling the missing postalcode values for Los Angeles
merged_df.loc[(merged_df['City'] =='Los Angeles') & (merged_df['PostalCode'].isnull()), 'PostalCode'] = '90001'
# we filled the missing postalcode values based on their city . By which we used the common postal code for each city

In [145]:
merged_df.isnull().sum()

OrderNumber         0
QuantityOrdered     0
PriceEach           0
OrderLineNumber     0
Sales               0
Status              0
QTR_ID              0
DAY_ID              0
MONTH_ID            0
YEAR_ID             0
ProductCode         0
CustomerName        0
Phone               0
AddressLine1        0
City                0
PostalCode          0
Country             0
Territory           0
ContactLastName     0
ContactFirstName    0
Branch              0
CostPrice           0
ProductLine         0
MSRP                0
dtype: int64

m

In [146]:
merged_df.head()

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,City,PostalCode,Country,Territory,ContactLastName,ContactFirstName,Branch,CostPrice,ProductLine,MSRP
0,10100,49,34.47,1,1689.03,Shipped,1,6,1,2022,...,Nashua,62005,USA,NA,Young,Valarie,Skates Road,22.4055,Vintage Cars,41
1,10110,48,34.47,5,1654.56,Shipped,1,18,3,2022,...,Manchester,EC2 5NT,UK,EMEA,Ashworth,Victoria,Skates Road,22.4055,Vintage Cars,41
2,10124,46,33.23,4,1528.58,Shipped,2,21,5,2022,...,Las Vegas,83030,USA,NA,King,Sue,Skates Road,21.5995,Vintage Cars,41
3,10149,26,38.98,9,1013.48,Shipped,3,12,9,2022,...,Brisbane,94217,USA,NA,Taylor,Sue,Skates Road,25.3370,Vintage Cars,41
4,10162,37,38.98,7,1442.26,Shipped,4,18,10,2022,...,San Francisco,94101,USA,NA,Brown,Julie,Skates Road,25.3370,Vintage Cars,41


In [147]:
merged_df.columns

Index(['OrderNumber', 'QuantityOrdered', 'PriceEach', 'OrderLineNumber',
       'Sales', 'Status', 'QTR_ID', 'DAY_ID', 'MONTH_ID', 'YEAR_ID',
       'ProductCode', 'CustomerName', 'Phone', 'AddressLine1', 'City',
       'PostalCode', 'Country', 'Territory', 'ContactLastName',
       'ContactFirstName', 'Branch', 'CostPrice', 'ProductLine', 'MSRP'],
      dtype='object')

In [148]:
# combinining the columns ColumnFirstName and ContactLastName to one column ContactName
merged_df['ContactName'] = merged_df['ContactFirstName'] + ' ' + merged_df['ContactLastName']
# droping the old columns
merged_df = merged_df.drop(columns=['ContactFirstName' , 'ContactLastName'])

In [149]:
merged_df.head()

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,DAY_ID,MONTH_ID,YEAR_ID,...,AddressLine1,City,PostalCode,Country,Territory,Branch,CostPrice,ProductLine,MSRP,ContactName
0,10100,49,34.47,1,1689.03,Shipped,1,6,1,2022,...,2304 Long Airport Avenue,Nashua,62005,USA,NA,Skates Road,22.4055,Vintage Cars,41,Valarie Young
1,10110,48,34.47,5,1654.56,Shipped,1,18,3,2022,...,Fauntleroy Circus,Manchester,EC2 5NT,UK,EMEA,Skates Road,22.4055,Vintage Cars,41,Victoria Ashworth
2,10124,46,33.23,4,1528.58,Shipped,2,21,5,2022,...,8489 Strong St.,Las Vegas,83030,USA,NA,Skates Road,21.5995,Vintage Cars,41,Sue King
3,10149,26,38.98,9,1013.48,Shipped,3,12,9,2022,...,2793 Furth Circle,Brisbane,94217,USA,NA,Skates Road,25.3370,Vintage Cars,41,Sue Taylor
4,10162,37,38.98,7,1442.26,Shipped,4,18,10,2022,...,7734 Strong St.,San Francisco,94101,USA,NA,Skates Road,25.3370,Vintage Cars,41,Julie Brown


In [151]:
# creating one date column
merged_df['Date'] = pd.to_datetime(merged_df[['YEAR_ID', 'MONTH_ID', 'DAY_ID']].astype(str).agg('-'.join,axis=1))
# droping the previous columns
merged_df = merged_df.drop(columns=['DAY_ID', 'MONTH_ID','YEAR_ID'])

In [152]:
merged_df

,OrderNumber,QuantityOrdered,PriceEach,OrderLineNumber,Sales,Status,QTR_ID,ProductCode,CustomerName,Phone,...,City,PostalCode,Country,Territory,Branch,CostPrice,ProductLine,MSRP,ContactName,Date
0,10100,49,34.47,1,1689.03,Shipped,1,S24_3969,Online Diecast Creations Co.,6035558647,...,Nashua,62005,USA,NA,Skates Road,22.4055,Vintage Cars,41,Valarie Young,2022-01-06
1,10110,48,34.47,5,1654.56,Shipped,1,S24_3969,AV Stores Co.,(171) 555-1555,...,Manchester,EC2 5NT,UK,EMEA,Skates Road,22.4055,Vintage Cars,41,Victoria Ashworth,2022-03-18
2,10124,46,33.23,4,1528.58,Shipped,2,S24_3969,Signal Gift Stores,7025551838,...,Las Vegas,83030,USA,NA,Skates Road,21.5995,Vintage Cars,41,Sue King,2022-05-21
3,10149,26,38.98,9,1013.48,Shipped,3,S24_3969,Signal Collectibles Ltd.,4155554312,...,Brisbane,94217,USA,NA,Skates Road,25.3370,Vintage Cars,41,Sue Taylor,2022-09-12
4,10162,37,38.98,7,1442.26,Shipped,4,S24_3969,Corporate Gift Ideas Co.,6505551386,...,San Francisco,94101,USA,NA,Skates Road,25.3370,Vintage Cars,41,Julie Brown,2022-10-18
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2818,10343,36,100.00,4,5848.92,Shipped,4,S18_1589,Reims Collectables,26.47.1555,...,Reims,51100,France,EMEA,Bergenville,65.0000,Classic Cars,124,Paul Henriot,2023-11-24
2819,10367,49,56.30,1,2758.70,Resolved,1,S18_1589,Toys4GrownUps.com,6265557265,...,Pasadena,90003,USA,NA,Skates Road,36.5950,Classic Cars,124,Julie Young,2024-01-12
2820,10378,34,42.64,5,1449.76,Shipped,1,S18_1589,Euro Shopping Channel,(91) 555 94 44,...,Madrid,28034,Spain,EMEA,Skates Road,27.7160,Classic Cars,124,Diego Freyre,2024-02-10
2821,10407,59,100.00,11,7048.14,On Hold,2,S18_1589,The Sharp Gifts Warehouse,4085553659,...,San Jose,94217,USA,NA,Geiselweg,65.0000,Classic Cars,124,Sue Frick,2024-04-22
